src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java

```diff
@@ -4,35 +4,198 @@
+	
+	/**
+	 * Velocity - this is better for programmers due to its precision
+	 * Percent - this is better for drivers because it feels more natural
+	 * TODO: Look at p.79 for a closed-loop walkthrough
+	 * TODO: Create a method for resetting calculated control values when
+	 * driver control (percent mode) begins; look through Talon methods first
+	 * TODO: Does Follower mode apply to speed as well as Control Mode?
+	 * @param leftSpeed rotations per minute
+	 */
+	public void driveLeftVelocity(double leftSpeed) {
+		if(leftSpeed >= RobotMap.VELOCITY_LOWER_SOFT_LIMIT && leftSpeed <= RobotMap.VELOCITY_UPPER_SOFT_LIMIT)
+		{
+			leftFrontTalon.changeControlMode(TalonControlMode.Speed);
+			leftFrontTalon.set(leftSpeed);
+		}
 	}
+	
+	public void driveLeftPercent(double leftSpeed) {
+		if(leftSpeed >= RobotMap.PERCENT_LOWER_SOFT_LIMIT && leftSpeed <= RobotMap.PERCENT_UPPER_SOFT_LIMIT )
+		{
+			leftFrontTalon.changeControlMode(TalonControlMode.PercentVbus);
+		
+			leftFrontTalon.set(leftSpeed);
+		}
+	}
+	
 
-	public void driveRight(double rightSpeed) {
-		rightFrontTalon.set(rightSpeed * RobotMap.DRIVETRAIN_FORWARD_RIGHT_CONSTANT);
-		rightBackTalon.set(rightSpeed * RobotMap.DRIVETRAIN_FORWARD_RIGHT_CONSTANT);
+	public void driveRightVelocity(double rightSpeed) {
+		if(rightSpeed >= RobotMap.VELOCITY_LOWER_SOFT_LIMIT && rightSpeed <= RobotMap.VELOCITY_UPPER_SOFT_LIMIT)
+		{
+			rightFrontTalon.changeControlMode(TalonControlMode.Speed);
+		
+			rightFrontTalon.set(rightSpeed);
+		}
+	}
+	
+	public void driveRightPercent(double rightSpeed) {
+		if(rightSpeed >= RobotMap.PERCENT_LOWER_SOFT_LIMIT && rightSpeed <= RobotMap.PERCENT_UPPER_SOFT_LIMIT )
+		{
+			rightFrontTalon.changeControlMode(TalonControlMode.PercentVbus);
+		
+			rightFrontTalon.set(rightSpeed);
+		}
 	}
 
+	
+	
+	/**
+	 * Gets the current position of the robot
+	 * Multiplied by the circumference of the wheel for scaling
+	 * @return
+	 */
+	public double getLeftPosition(){
+		return leftFrontTalon.getPosition() * RobotMap.DRIVETRAIN_WHEEL_CIRCUMFERENCE;
+	}
+	
+	public double getRightPosition(){
+		return rightFrontTalon.getPosition() * RobotMap.DRIVETRAIN_WHEEL_CIRCUMFERENCE;
+	}
+	
+	/**
+	 * Methods that get/set the left and right positions of where we want to be
+	 * setPosition zeros out the encoder data, so that we start at 0
+	 */
+	
+	public void setLeftDesiredPosition(double goalDistance){
+		leftFrontTalon.setPosition(0);
+		this.goalDistance = goalDistance;
+	}
+	
+	public void setRightDesiredPosition(double goalDistance){
+		rightFrontTalon.setPosition(0);
+		this.goalDistance = goalDistance;
+	}
+	
+	public double getLeftDesiredDistance()
+	{
+		return goalDistance;
+	}
+	
+	public double getRightDesiredDistance()
+	{
+		return goalDistance;
+	}
+	
+	/**
+	 * Methods that calculate the error in distance from where we are to where we want to be
+	 */
+	public double calculateLeftDistanceError(){
```

Comment:
Minor naming quibble: Are we "calculating" anything, or are we just getting it? (I'm open to be convinced either way)